core/src/muxing.rs

```diff
@@ -51,178 +50,61 @@
-    /// call this method may be notified.
-    ///
-    /// Calling `write_substream` does not guarantee that data will arrive to the remote. To
-    /// ensure that, you should call `flush_substream`.
-    ///
-    /// It is incorrect to call this method on a substream if you called `shutdown_substream` on
-    /// this substream earlier.
-    fn write_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-        buf: &[u8],
-    ) -> Poll<Result<usize, Self::Error>>;
-
-    /// Flushes a substream. The behaviour is the same as `futures::AsyncWrite::poll_flush`.
-    ///
-    /// After this method has been called, data written earlier on the substream is guaranteed to
-    /// be received by the remote.
-    ///
-    /// If `Pending` is returned, then the current task will be notified once the substream
-    /// is ready to be read. For each individual substream, only the latest task that was used to
-    /// call this method may be notified.
-    ///
-    /// > **Note**: This method may be implemented as a call to `flush_all`.
-    fn flush_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>>;
-
-    /// Attempts to shut down the writing side of a substream. The behaviour is similar to
-    /// `AsyncWrite::poll_close`.
-    ///
-    /// Contrary to `AsyncWrite::poll_close`, shutting down a substream does not imply
-    /// `flush_substream`. If you want to make sure that the remote is immediately informed about
-    /// the shutdown, use `flush_substream` or `flush_all`.
-    ///
-    /// After this method has been called, you should no longer attempt to write to this substream.
-    ///
-    /// An error can be generated if the connection has been closed, or if a protocol misbehaviour
-    /// happened.
-    fn shutdown_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>>;
-
-    /// Destroys a substream.
-    fn destroy_substream(&self, s: Self::Substream);
-
-    /// Closes this `StreamMuxer`.
+    /// Closes this [`StreamMuxer`].
     ///
     /// After this has returned `Poll::Ready(Ok(()))`, the muxer has become useless. All
     /// subsequent reads must return either `EOF` or an error. All subsequent writes, shutdowns,
     /// or polls must generate an error or be ignored.
     ///
-    /// Calling this method implies `flush_all`.
-    ///
     /// > **Note**: You are encouraged to call this method and wait for it to return `Ready`, so
     /// >           that the remote is properly informed of the shutdown. However, apart from
     /// >           properly informing the remote, there is no difference between this and
     /// >           immediately dropping the muxer.
-    fn poll_close(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
```

Comment:
```suggestion
    fn poll_close(mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
```
Why not move ownership if the muxer should not be used afterwards anyway?

Reply:
It is still a poll function that may return Pending and needs to be polled again!

Reply:
Ahh yes of course you're right. Somehow that completely slipped my mind.